Feature: Svelte Wrapper

Zakrok09 opened the pull request:
## Svelte Wrapper of uPlot 

closes #29

This pull request: 
- adds a Svelte component library with a single uPlot wrapper component that follows the structure and behavior of its React and Vue counterparts 
- updates the README to reflect on that 
- provides an example use of the component in `svelte/src/routes/+page.svelte`

### Demonstration 

https://github.com/skalinichev/uplot-wrappers/assets/31936449/7ff78e43-1939-4b83-b206-a41045549ec8

In the video, the log paragraph updates with words "Created" and "Destroyed". Those are done by functions passed as `onCreate` and `onDestroy` props to the component. Updating or adding to the data and updating the width  does not cause a "full-reload" - creation and deletion. However, changing the colour of the series does. Toggling the component also causes a "full-reload".


## Problems

I had trouble integration with the commons package as it was private and I could not install it to use its method for checking state of the options changes. Therefore, I copy-pasted the function and left the importing commented out. Uncomment if the issue has been resolved and remove its re-definition in the UplotSvelte.svelte component file.

skalinichev commented on svelte/package.json:
```diff
@@ -0,0 +1,69 @@
+		"Wrapper"
+	],
+	"scripts": {
+		"dev": "vite dev",
```
Can we use webpack for that? The idea is to use the same tools for all wrappers. While it may be not that convenient, it keeps the code-base consistent. It would be really difficult for me to maintain it if every wrapper uses it's own tools(
Please, see if you can do it, otherwise i'll have to do it myself later on.

Zakrok09 replied:
I attempted to do it myself, I believe it is very much possible, the only issue is that I am having trouble with the existing webpack.configuration. I leave it as vite for now because I am afraid I cannot do it myself.

I will instead link articles and packages that I believe would be very helpful:

- [Svelte + Webpack 5 article](https://codechips.me/svelte-and-webpack-5/)
- [svelte-loader](https://github.com/sveltejs/svelte-loader)
- [svelte-preprocess](https://github.com/sveltejs/svelte-preprocess)

skalinichev commented on svelte/package.json:
```diff
@@ -0,0 +1,69 @@
+	},
+	"files": [
+		"dist",
+		"!dist/**/*.test.*",
```
Please, clean this file up a bit. Seems like it was auto-generated and has some not-used entries. Try to keep it simple by leaving only the required ones.

skalinichev commented on svelte/package.json:
```diff
@@ -0,0 +1,69 @@
+		"svelte": "^4.2.7",
+		"svelte-check": "^3.6.0",
+		"tslib": "^2.4.1",
+		"typescript": "^5.0.0",
```
The root project already uses typescript, please try to reuse it

skalinichev commented on svelte/package.json:
```diff
@@ -0,0 +1,69 @@
+		"publint": "^0.1.9",
+		"svelte": "^4.2.7",
+		"svelte-check": "^3.6.0",
+		"tslib": "^2.4.1",
```
Do we really need it, what it being used for?

skalinichev commented on svelte/package.json:
```diff
@@ -0,0 +1,69 @@
+		"@sveltejs/kit": "^2.0.0",
+		"@sveltejs/package": "^2.0.0",
+		"@sveltejs/vite-plugin-svelte": "^3.0.0",
+		"publint": "^0.1.9",
```
I'm not expecting there to be many changes in the package.json file. Can we drop it? The idea is to reduce the dependencies to a bare minimum. If we decide to add it in the future it should be placed in the root directory anyway.

skalinichev commented on svelte/src/app.d.ts:
```diff
@@ -0,0 +1,13 @@
+// See https://kit.svelte.dev/docs/types#app
```
I followed the link, but still don't understand where it's being used in the application. Is it required for the library or for the sample application?  In the first case we should add it into the exported public interface in the later move it into the example itself.

Zakrok09 replied:
This file would be for the sample application example. The library itself does not use it. Vite could use the file when bundling into a package

skalinichev commented on svelte/src/lib/UplotSvelte.svelte:
Is it possible to keep a simple one-level project structure? While it may not be the recommended way to write svelte projects, we want to keep the code-base consistent across the wrappers. You can also move everything related to the example to example folder. The code is not that complex anyway, there is not much point in doing it.

Zakrok09 replied:
![image](https://github.com/skalinichev/uplot-wrappers/assets/31936449/19634061-377a-45d1-b43d-e44366bd7f21)

Those will be the only files exported and present in the package once it gets exported. I tried to just copy paste them into the `/svelte`. I had difficulties pointing them in the package.json. I will try again shortly.

skalinichev replied:
I was talking more about project structure: lib, routes, static. It's a very simple application and it seems like an overkill to me to create such a complex project structure for it. Would it be possible to keep just 2 files - one is the library wrapper, another one is an usage example inside the svelte directory?

Zakrok09 replied:
Certainly, if one instead just puts the UplotSvelte.svelte file in the `/svelte` and the code from `+page.svelte` and rename it into an `example.svelte` it should be great. Unfortunately, I have to keep the structure because it's what the svelte packager uses to bundle everything to a package.

skalinichev commented on svelte/src/lib/UplotSvelte.svelte:
```diff
@@ -0,0 +1,104 @@
+<script lang="ts">
+    import 'uplot/dist/uPlot.min.css';
+    import type uPlot from 'uplot';
+    import {onDestroy, onMount} from "svelte";
```
Please, run prettier on the code-base, it's doesn't seem to be formatted to me. You can use the .prettierrc config in the root directory.

skalinichev commented on svelte/src/lib/UplotSvelte.svelte:
```diff
@@ -0,0 +1,104 @@
+    import 'uplot/dist/uPlot.min.css';
+    import type uPlot from 'uplot';
+    import {onDestroy, onMount} from "svelte";
+    // import {optionsUpdateState} from "uplot-wrappers-common";
```
The project uses yarn workspaces. You need to add a new entry to the root package.json file in the workspaces section to avoid copying the library

Zakrok09 replied:
Apologies, I was not familiar with yarn and its workspaces feature. This now works and the problem isn't present. 

As a side question: would you say that yarn is more preferable to npm generally? Thanks!

skalinichev replied:
Can't say for everyone, but the recent project I worked with all using yarn.
The reason is that yarn provides more features like resolutions, workspaces etc. Having said that it seems like npm started to catch up with yarn a bit in terms of functionality, so it may not be an issue anymore.

skalinichev commented on svelte/src/lib/UplotSvelte.svelte:
```diff
@@ -0,0 +1,104 @@
+    export let onCreate: (chart: uPlot) => void = () => {};
+    export let resetScales:boolean = true;
+
+    let chart: uPlot | null = null;
```
Wait, how does it even work, is it a global variable or something? What will happen if one creates 2 instances of the UplotSvelte component? Will the chart from the second instance override the first one or will it somehow create a member variable for that during the runtime? Sorry if it's a stupid question, just want to double check how it actually works.

skalinichev commented on svelte/src/lib/UplotSvelte.svelte:
```diff
@@ -0,0 +1,104 @@
+    };
+
+    onMount(async () => {
+        const uplotModule = await import('uplot');
```
Why do we have to import it dynamically? Please, import everything statically, there is no need for it right now.

Zakrok09 replied:
There is unfortunately I problem I ended up with when trying to import it statically. Uplot requests access to the DOM. In svelte it could be that the code you are running can be still in the server and the document still not loaded. By using `onMount` you would ensure that the code inside is ran when the document object is accessible. 

To make it accessible to Svelte users (and not only SvelteKit) it is not possible to use SvelteKit's elegant way of checking this - [docs entry for best practices](https://kit.svelte.dev/docs/packaging#best-practices).

The other alternative would be to use the package [esm-env](https://github.com/benmccann/esm-env). However, I am not sure if you'd agree with adding more dependencies.

skalinichev replied:
Hmm, that is strange. uPlot shouldn't access DOM before creating an instance of it. So I though it should work just fine... Could you share the error that you have encountered there?

Zakrok09 replied:
```
document is not defined

ReferenceError: document is not defined
    at Object.<anonymous> (/home/.../Projects/uplot-wrappers/node_modules/uplot/dist/uPlot.cjs.js:487:13)
    at Module._compile (node:internal/modules/cjs/loader:1378:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1437:10)
    at Module.load (node:internal/modules/cjs/loader:1212:32)
    at Module._load (node:internal/modules/cjs/loader:1028:12)
    at cjsLoader (node:internal/modules/esm/translators:359:17)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:308:7)
    at ModuleJob.run (node:internal/modules/esm/module_job:218:25)
    at async ModuleLoader.import (node:internal/modules/esm/loader:323:24)
    at async nodeImport (file:///home/.../Projects/uplot-wrappers/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:55096:17
    ```
    
    Essentially `document` not found. I believe this is a common issue for Svelte + uPlot and one has to use either workaround for it.

skalinichev reviewed (changes requested):
There are some minor issues with the integration into the existing code base, apart from that looks good to me!
Thanks for the hard work!

Zakrok09 commented on svelte/.npmrc:
```diff
@@ -0,0 +1 @@
+engine-strict=true
```
I believe this was generated by the `@sveltejs/package` component library generator. 

It is an option when generating Svelte projects to choose it to be component library.

skalinichev replied:
Read up on it a bit. It seems to be an option to enforce node version for dependencies. In our case it's not really needed, engines.node entry in package.json should be enough. So we can safely remove it.

Zakrok09 commented on svelte/README.md:
Apologies, I must've forgot

Zakrok09 commented on svelte/package.json:
```diff
@@ -0,0 +1,69 @@
+		"SvelteKit",
+		"Wrapper"
+	],
+	"scripts": {
```
Svelte uses Vite and not webpack. I hope this won't be an issue. I will try to move the scripts but I assume it would require to also put the vite.config.js in the root.

Zakrok09 commented:
Comments on recent changes:

- I kept Typescript because the bundler for svelte package requires typescript 5+. 
- I linked the workspaces and it works. Unfortunately, I am running in an issue where when I bundle the package to a tar.gz it breaks when I upload it another package because the `uplot-wrappers-common` package cannot be imported since its private. This somehow was not an issue earlier today but now I cannot seem to fix it. I assume that by using webpack to bundle the package this will get resolved. 
- I prettified the code with prettier
- I cleaned up some packages and removed unnecessary scripts


As I said, I attempted to use webpack to build it and I am more than positive that it won't be difficult. I just had difficulties with the already existing webpack configuration and how to apply the changes there. I am leaving helpful articles and packages to help you with it. I believe it would be easier for you since you know how to operate the existing script.

- [Svelte + Webpack 5 article](https://codechips.me/svelte-and-webpack-5/)
- [svelte-loader](https://github.com/sveltejs/svelte-loader)
- [svelte-preprocess](https://github.com/sveltejs/svelte-preprocess)

skalinichev commented:
I think we can merge it as is. I'll try to do a better integration into the existing code-base these or the following weekends and release it shortly. Again many thanks for your contribution!
